x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java

```diff
@@ -972,22 +996,38 @@ private static Expression processScalarFunction(EsqlScalarFunction f, EsqlFuncti
             boolean childrenChanged = false;
             DataType targetDataType = DataType.NULL;
             Expression arg;
+            DataType targetNumericType = DataType.NULL;
+            boolean castNumericArgs = false;
             for (int i = 0; i < args.size(); i++) {
                 arg = args.get(i);
-                if (arg.resolved() && arg.dataType() == KEYWORD && arg.foldable() && ((arg instanceof EsqlScalarFunction) == false)) {
-                    if (i < targetDataTypes.size()) {
-                        targetDataType = targetDataTypes.get(i);
-                    }
-                    if (targetDataType != DataType.NULL && targetDataType != DataType.UNSUPPORTED) {
-                        Expression e = castStringLiteral(arg, targetDataType);
-                        childrenChanged = true;
-                        newChildren.add(e);
-                        continue;
+                if (arg.resolved()) {
+                    var dataType = arg.dataType();
+                    if (dataType == KEYWORD) {
+                        if (arg.foldable() && ((arg instanceof EsqlScalarFunction) == false)) {
+                            if (i < targetDataTypes.size()) {
+                                targetDataType = targetDataTypes.get(i);
+                            }
+                            if (targetDataType != DataType.NULL && targetDataType != DataType.UNSUPPORTED) {
+                                Expression e = castStringLiteral(arg, targetDataType);
+                                childrenChanged = true;
+                                newChildren.add(e);
+                                continue;
+                            }
+                        }
+                    } else if (dataType.isNumeric() && canCastMixedNumericTypes(f)) {
```

Comment:
I'm trying to do both string literal casting and numeric casting within one loop of the input arguments. `COALESCE` could be extracted to its own method like `IN`, however if `COALESCE` is in its own method, we will either need to make a copy of `processScalarFunction` inside `processCoalesce`, or have `processCoalesce` call `processScalarFunction` and make a separate loop to do numeric casting. Although `IN` extends `EsqlScalarFunction`, it is different from other `EsqlScalarFunction`s, as it is not registered in `EsqlFunctionRegistry`,   `processScalarFunction` doesn't quite work for `IN`, this is the main reason that `IN` has its own method.

Reply:
> if COALESCE is in its own method, we will either need to make a copy of `processScalarFunction` inside `processCoalesce`

I'm not sure we'd need to duplicate that much code. The numeric casting (`targetNumericType` & co.) and string casting (`targetDataTypes` & co.) have nothing in common, except iterating over function's arguments list. Also, out of the loop, children updating (on `childrenChanged`) never happens at the same time with numerical casting (`castNumericArgs`). 
Then checking if `canCastMixedNumericTypes(f)` at every loop iteration, as if it can change, looks a bit suspicious. I'd maybe do that only when `targetNumericType` is still `null`.